crates/matrix-sdk/src/authentication/oidc/backend/mock.rs

```diff
@@ -240,4 +240,33 @@ impl OidcBackend for MockImpl {
             })
         }
     }
+
+    #[cfg(all(feature = "e2e-encryption", not(target_arch = "wasm32")))]
+    async fn request_device_authorization(
+        &self,
+        _device_authorization_endpoint: Url,
+        _client_id: oauth2::ClientId,
+        _scopes: Vec<oauth2::Scope>,
+    ) -> Result<
+        oauth2::StandardDeviceAuthorizationResponse,
+        oauth2::basic::BasicRequestTokenError<oauth2::HttpClientError<reqwest::Error>>,
+    > {
+        unimplemented!()
+    }
+
+    #[cfg(all(feature = "e2e-encryption", not(target_arch = "wasm32")))]
+    async fn exchange_device_code(
+        &self,
+        _token_endpoint: Url,
+        _client_id: oauth2::ClientId,
+        _device_authorization_response: &oauth2::StandardDeviceAuthorizationResponse,
+    ) -> Result<
+        OidcSessionTokens,
+        oauth2::RequestTokenError<
+            oauth2::HttpClientError<reqwest::Error>,
+            oauth2::DeviceCodeErrorResponse,
+        >,
+    > {
+        unimplemented!()
+    }
```

Comment:
To be honest, I don't really like this backend trait and that we're using it to test the OIDC submodule.

I wish we properly mocked things instead and added a `oidc()` subtype to the `MatrixMockServer`.

I know it's annoying to mock the OIDC stuff, but if the server changes under our feet we don't have any proof that it did so. Furthermore we don't have any integration tests that use MAS either.

I guess it's still a good step to centralize the logic instead of having this additional OIDC client thing, and then we can mock things using `wiremock` and `MatrixMockServer` and eventually get rid of this mock backend and trait.

Reply:
I had planned to slightly refactor the tests before changing to the oauth2 crate, to reduce duplication and simplify the migration. Maybe I can do this instead then.